### PR TITLE
proxy: remove impersonate headers for kubernetes

### PR DESCRIPTION
**Pull request by calebdoxsey**
## Summary
When Kubernetes is being proxied we want to remove the `Impersonate-...` headers. This PR does that using a lua script.

It also sets the first group to be a replace, and then subsequent groups to be an append.

## Related issues

**Checklist**:
- [ ] add related issues
- [ ] updated docs
- [x] updated unit tests
- [ ] updated UPGRADING.md
- [x] ready for review

### CODE REVIEW

File: internal/controlplane/luascripts/remove-impersonate-headers.lua [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    return str:sub(1, #start) == start
+end
+
+function envoy_on_request(request_handle)
+    local headers = request_handle:headers()
+    local metadata = request_handle:metadata()
+
+    local remove_impersonate_headers = metadata:get("remove_impersonate_headers")
+    if remove_impersonate_headers then
+        local to_remove = {}
```
Comment (calebdoxsey) [inline internal/controlplane/luascripts/remove-impersonate-headers.lua:11]: according to the docs we aren't allowed to modify the header collection while looping over it

File: authorize/check_response.go [hunk 2]
```diff
@@ -154,15 +154,15 @@ func (a *Authorize) redirectResponse(in *envoy_service_auth_v2.CheckRequest) *en
...
 		requestHeaders = append(requestHeaders,
 			mkHeader("Authorization", "Bearer "+reply.MatchingPolicy.KubernetesServiceAccountToken, false))
 
 		if reply.UserEmail != "" {
 			requestHeaders = append(requestHeaders, mkHeader("Impersonate-User", reply.UserEmail, false))
 		}
-		for _, group := range reply.UserGroups {
-			requestHeaders = append(requestHeaders, mkHeader("Impersonate-Group", group, true))
+		for i, group := range reply.UserGroups {
+			requestHeaders = append(requestHeaders, mkHeader("Impersonate-Group", group, i > 0))
```
Comment (desimone) [inline authorize/check_response.go:165]: clever

File: internal/controlplane/luascripts/remove-impersonate-headers.lua [hunk 3]
```diff
@@ -0,0 +1,25 @@
+local function starts_with(str, start)
+    return str:sub(1, #start) == start
```
Comment (desimone) [inline internal/controlplane/luascripts/remove-impersonate-headers.lua:2]: 1_iJhMo_FJ6ka70UU1t_h1eQ

File: internal/controlplane/luascripts/remove-impersonate-headers.lua [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+            end
+        end
+
+        for k, v in pairs(to_remove) do
+            headers:remove(v)
+        end
+    end
+end
+
+function envoy_on_response(response_handle)
```
Comment (desimone) [inline internal/controlplane/luascripts/remove-impersonate-headers.lua:24]: Is this required to fulfill the interface?

File: internal/frontend/statik/statik.go [hunk 5]
```diff
@@ -10,6 +10,6 @@ import (
...
...x00\x00\x00\x00\x00\xb4\x81\xc3\x08\x00\x00img/account_circle-24px.svgUT\x05\x00\x01\xa9\xe1\xc2^PK\x01\x02\x14\x03\x14\x00\x08\x00\x08\x00r\x9b\xb2P\xfc\xc6x\x8f\xb7\x00\x00\x00\xf9\x00\x00\x00\x12\x00	\x00\x00\x00\x00\x00\x00\x00\x00\x00\xb4\x81\n\n\x00\x00img/error-24px.svgUT\x05\x00\x01\xa9\xe1\xc2^PK\x01\x02\x14\x03\x14\x00\x08\x00\x08\x00r\x9b\xb2PK\xfe\x8b#M\x03\x00\x00d\x08\x00\x00\x10\x00	\x00\x00\x00\x00\x00\x00\x00\x00\x00\xb4\x81\n\x0b\x00\x00img/pomerium.svgUT\x05\x00\x01\xa9\xe1\xc2^PK\x01\x02\x14\x03\x14\x00\x08\x00\x08\x00\xcc\xb1\xfcP\xff9\xbd(rt\x00\x00\x12\xa0\x00\x00\x1a\x00	\x00\x00\x00\x00\x00\x00\x00\x00\x00\xb4\x81\x9e\x0e\x00\x00img/pomerium_circle_96.svgUT\x05\x00\x01@\xa3 _PK\x01\x02\x14\x03\x14\x00\x08\x00\x08\x00r\x9b\xb2Puq\x02\xd2d\x01\x00\x00\x9e\x02\x00\x00#\x00	\x00\x00\x00\x00\x00\x00\x00\x00\x00\xb4\x81a\x83\x00\x00img/supervised_user_circle-24px.svgUT\x05\x00\x01\xa9\xe1\xc2^PK\x01\x02\x14\x03\x14\x00\x08\x00\x08\x00r\x9b\xb2PL\xbb\xd3^\xd3\x05\x00\x00^\x12\x00\x00\x0e\x00	\x00\x00\x00\x00\x00\x00\x00\x00\x00\xb4\x81\x1f\x85\x00\x00style/main.cssUT\x05\x00\x01\xa9\xe1\xc2^PK\x05\x06\x00\x00\x00\x00	\x00	\x00\xb2\x02\x00\x007\x8b\x00\x00\x00\x00"
```
Comment (desimone) [inline internal/frontend/statik/statik.go:13]: Accident?

**Review by desimone — APPROVED** [review]
LGTM; just a nit, question, and meme.

**Review by travisgroth — APPROVED** [review]
LGTM pending @desimone's comments.

**Comment by desimone** [comment]
\cc @whs

File: internal/controlplane/luascripts/remove-impersonate-headers.lua (hunk 4 above)
Comment (calebdoxsey) [reply internal/controlplane/luascripts/remove-impersonate-headers.lua:24]: yes, it has an annoying log message if removed

**Comment by calebdoxsey** [comment]
This will not fix the comma-delimited group issue. That's how envoy constructs multiple headers and I'm not sure what we can do to fix it.

**Comment by travisgroth** [comment]
> This will not fix the comma-delimited group issue. That's how envoy constructs multiple headers and I'm not sure what we can do to fix it.

👍 

I'm looking into the group permissions not working separately.  kubectl also uses a single comma separated group header for impersonation, so I don't think that's a fundamental problem.  It's something more nuanced about the header.

**Pull request merged**